subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c

```diff
@@ -375,35 +412,80 @@ int img_mgmt_set_next_boot_slot(int slot, bool confirm)
 	const struct flash_area *fa;
 	int area_id = img_mgmt_flash_area_id(slot);
 	int rc = 0;
-	bool active = (slot == img_mgmt_active_slot(img_mgmt_active_image()));
-
-#if defined(CONFIG_MCUMGR_GRP_IMG_STATUS_HOOKS)
-	int32_t ret_rc;
-	uint16_t ret_group;
+	/* image the requested slot is defined within */
+	int image = img_mgmt_slot_to_image(slot);
+	/* active_slot is slot that is considered active/primary/executing
+	 * for a given image.
+	 */
+	int active_slot = img_mgmt_active_slot(image);
+	enum img_mgmt_next_boot_type type;
+	int next_boot_slot = img_mgmt_get_next_boot_slot(image, &type);
+
+	LOG_DBG("(%d, %s)", slot, confirm ? "confirm" : "test");
+	LOG_DBG("aimg = %d, img = %d, aslot = %d, slot = %d, nbs = %d",
+		img_mgmt_active_image(), image, active_slot, slot, next_boot_slot);
+
+	/* MCUmgr should not allow to confirm non-active image slots to prevent
+	 * confiming something that might not have been verified to actually be bootable
+	 * or have stuck in primary slot of other image. Unfortunately there was
+	 * a bug in logic that always allowed to confirm secondary slot of any
+	 * image. Now the behaviour is controlled via Kconfig options.
+	 */
+#ifndef CONFIG_MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_ANY
+	if (confirm && image != img_mgmt_active_image() &&
+	    (!IS_ENABLED(CONFIG_MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_SECONDARY) ||
+	     slot == active_slot)) {
+		LOG_DBG("Not allowed to confirm non-active images");
+		return IMG_MGMT_RET_RC_INVALID_SLOT;
```

Comment:
So generally I should drop the entire check?

Reply:
Previously we did not allow to confirm non-running image.

Reply:
@nordicjm  Should I completely drop the `CONFIG_MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_SECONDARY` and `CONFIG_MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_ANY` ?

Reply:
There doesn't look to be an hook in this code which allows the change to be declined, that would be required to drop this, in my mind this should be entirely user/application driven, if a user wants to allow or deny confirming slots for specific reasons, they should decide on those rules because they will probably be application and mcuboot operating mode specific

Reply:
Leave this in but update the error code, hook version will be a "version 2: in the future!" feature, probably when someone asks for it